File: mahjong/shanten.py
from __future__ import annotations

from functools import lru_cache


def normal_shanten(counts: tuple[int, ...] | list[int]) -> int:
    """Return shanten number for normal hands only.

    Winning normal hands return -1, tenpai returns 0, and iishanten returns 1.
    Chiitoitsu and kokushi are intentionally excluded so they can be added as
    separate strategies later.
    """
    counts_tuple = tuple(counts)
    if len(counts_tuple) != 34:
        raise ValueError("normal_shanten expects 34 tile counts.")

    best = 8

    for pair_index in [None] + [index for index, count in enumerate(counts_tuple) if count >= 2]:
        working = list(counts_tuple)
        has_pair = 0
        if pair_index is not None:
            working[pair_index] -= 2
            has_pair = 1

        for melds, taatsu in _block_options(tuple(working)):
            capped_taatsu = min(taatsu, 4 - melds)
            best = min(best, 8 - (2 * melds) - capped_taatsu - has_pair)

    return best


@lru_cache(maxsize=None)
def _block_options(counts: tuple[int, ...]) -> frozenset[tuple[int, int]]:
    first = next((index for index, count in enumerate(counts) if count), None)
    if first is None:
        return frozenset({(0, 0)})

    results: set[tuple[int, int]] = set()

    def add_branch(next_counts: list[int], meld_delta: int = 0, taatsu_delta: int = 0) -> None:
        for melds, taatsu in _block_options(tuple(next_counts)):
            results.add((melds + meld_delta, taatsu + taatsu_delta))

    # Ignore this tile as an isolated tile.
    skipped = list(counts)
    skipped[first] -= 1
    add_branch(skipped)

    if counts[first] >= 3:
        triplet = list(counts)
        triplet[first] -= 3
        add_branch(triplet, meld_delta=1)

    if _can_start_sequence(first) and counts[first + 1] and counts[first + 2]:
        sequence = list(counts)
        sequence[first] -= 1
        sequence[first + 1] -= 1
        sequence[first + 2] -= 1
        add_branch(sequence, meld_delta=1)

    if counts[first] >= 2:
        pair_taatsu = list(counts)
        pair_taatsu[first] -= 2
        add_branch(pair_taatsu, taatsu_delta=1)

    if _can_connect(first, 1) and counts[first + 1]:
        adjacent = list(counts)
        adjacent[first] -= 1
        adjacent[first + 1] -= 1
        add_branch(adjacent, taatsu_delta=1)

    if _can_connect(first, 2) and counts[first + 2]:
        closed_wait = list(counts)
        closed_wait[first] -= 1
        closed_wait[first + 2] -= 1
        add_branch(closed_wait, taatsu_delta=1)

    return frozenset(results)


def _can_start_sequence(index: int) -> bool:
    return index < 27 and index % 9 <= 6


def _can_connect(index: int, distance: int) -> bool:
    return index < 27 and (index // 9) == ((index + distance) // 9) and (index % 9) + distance <= 8

File: packages/mahjong-core/src/shanten.ts
import type { Counts34 } from "./tiles";
import { LruCache, incrementSimulationCounter } from "./performance";

const blockCache = new LruCache<string, readonly number[]>(50_000);
const shantenCache = new LruCache<string, number>(50_000);
const LOCAL_TAATSU_BASE = 8;
const COMBINED_TAATSU_BASE = 16;

export function normalShanten(counts: Counts34): number {
  return normalShantenWithOpenMelds(counts, 0);
}

export function normalShantenWithOpenMelds(counts: Counts34, openMeldCount: number): number {
  if (counts.length !== 34) {
    throw new Error("normalShantenWithOpenMelds expects 34 tile counts.");
  }
  if (!Number.isInteger(openMeldCount) || openMeldCount < 0 || openMeldCount > 4) {
    throw new Error("openMeldCount must be an integer from 0 to 4.");
  }
  const cacheKey = `${openMeldCount}|${counts.join(",")}`;
  const cached = shantenCache.get(cacheKey);
  if (cached != null) return cached;
  incrementSimulationCounter("shantenCalculationCount");

  let best = 8 - openMeldCount * 2;
  const pairOptions: Array<number | null> = [null];
  counts.forEach((count, index) => {
    if (count >= 2) pairOptions.push(index);
  });

  for (const pairIndex of pairOptions) {
    const working = counts.slice();
    let hasPair = 0;
    if (pairIndex != null) {
      working[pairIndex] -= 2;
      hasPair = 1;
    }

    for (const option of blockOptions(working)) {
      const melds = option >> 4;
      const taatsu = option & 15;
      const totalMelds = melds + openMeldCount;
      if (totalMelds > 4) continue;
      const cappedTaatsu = Math.min(taatsu, 4 - totalMelds);
      best = Math.min(best, 8 - 2 * totalMelds - cappedTaatsu - hasPair);
    }
  }

  shantenCache.set(cacheKey, best);
  return best;
}

function blockOptions(counts: Counts34): readonly number[] {
  let combined: readonly number[] = [0];
  for (const [start, length, allowSequences] of [
    [0, 9, true],
    [9, 9, true],
    [18, 9, true],
    [27, 7, false],
  ] as const) {
    const local = localBlockOptions(counts.slice(start, start + length), allowSequences);
    const next = new Set<number>();
    for (const current of combined) {
      const currentMelds = current >> 4;
      const currentTaatsu = current & 15;
      for (const option of local) {
        const melds = currentMelds + (option >> 3);
        const taatsu = currentTaatsu + (option & 7);
        next.add(melds * COMBINED_TAATSU_BASE + taatsu);
      }
    }
    combined = [...next];
  }
  return combined;
}

function localBlockOptions(counts: number[], allowSequences: boolean): readonly number[] {
  const key = `${allowSequences ? "s" : "h"}${encodeBaseFive(counts)}`;
  const cached = blockCache.get(key);
  if (cached) return cached;
  const first = counts.findIndex((count) => count > 0);
  if (first < 0) {
    const empty = [0] as const;
    blockCache.set(key, empty);
    return empty;
  }

  const results = new Set<number>();
  const addBranch = (nextCounts: number[], meldDelta = 0, taatsuDelta = 0) => {
    for (const option of localBlockOptions(nextCounts, allowSequences)) {
      results.add(
        ((option >> 3) + meldDelta) * LOCAL_TAATSU_BASE
          + (option & 7)
          + taatsuDelta,
      );
    }
  };

  const skipped = counts.slice();
  skipped[first] -= 1;
  addBranch(skipped);

  if (counts[first]! >= 3) {
    const triplet = counts.slice();
    triplet[first] -= 3;
    addBranch(triplet, 1);
  }

  if (allowSequences && first <= 6 && counts[first + 1]! > 0 && counts[first + 2]! > 0) {
    const sequence = counts.slice();
    sequence[first] -= 1;
    sequence[first + 1] -= 1;
    sequence[first + 2] -= 1;
    addBranch(sequence, 1);
  }

  if (counts[first]! >= 2) {
    const pairTaatsu = counts.slice();
    pairTaatsu[first] -= 2;
    addBranch(pairTaatsu, 0, 1);
  }

  if (allowSequences && first <= 7 && counts[first + 1]! > 0) {
    const adjacent = counts.slice();
    adjacent[first] -= 1;
    adjacent[first + 1] -= 1;
    addBranch(adjacent, 0, 1);
  }

  if (allowSequences && first <= 6 && counts[first + 2]! > 0) {
    const closedWait = counts.slice();
    closedWait[first] -= 1;
    closedWait[first + 2] -= 1;
    addBranch(closedWait, 0, 1);
  }

  const result = [...results];
  blockCache.set(key, result);
  return result;
}

function encodeBaseFive(counts: readonly number[]): number {
  let encoded = 0;
  for (const count of counts) encoded = encoded * 5 + count;
  return encoded;
}
